File: cat.py
#!/usr/bin/env python3
import sys
import os

def cat_file(f):
    try:
        # lese/schreibe im Binary-Mode, so verhält sich cat korrekt auch bei Binärdateien
        while True:
            chunk = f.read(8192)
            if not chunk:
                break
            sys.stdout.buffer.write(chunk)
    except BrokenPipeError:
        # z.B. wenn Ausgabe in `head` gepiped wurde und vorzeitig schließt
        sys.exit(0)

def main():
    args = sys.argv[1:]
    if not args:
        # keine Dateien => von stdin lesen
        cat_file(sys.stdin.buffer)
    else:
        for name in args:
            if name == "-":
                cat_file(sys.stdin.buffer)
            else:
                try:
                    with open(name, "rb") as f:
                        cat_file(f)
                except OSError as e:
                    # Fehler mimimale Ausgabe an stderr wie /bin/cat
                    sys.stderr.write(f"cat: {name}: {e.strerror}\n")

if __name__ == "__main__":
    main()

File: cat.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
    FILE *f;
    char buf[8192];
    size_t n;

    if (argc == 1) {
        // Keine Argumente -> stdin
        while ((n = fread(buf, 1, sizeof(buf), stdin)) > 0) {
            fwrite(buf, 1, n, stdout);
        }
    } else {
        for (int i = 1; i < argc; i++) {
            if (argv[i][0] == '-' && argv[i][1] == '\0') {
                f = stdin;
            } else {
                f = fopen(argv[i], "rb");
                if (!f) {
                    fprintf(stderr, "cat: %s: cannot open\n", argv[i]);
                    continue;
                }
            }

            while ((n = fread(buf, 1, sizeof(buf), f)) > 0) {
                fwrite(buf, 1, n, stdout);
            }

            if (f != stdin) fclose(f);
        }
    }

    return 0;
}
